cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/AuthorizationArtifactRepository.java

```diff
@@ -0,0 +1,299 @@
+import co.cask.cdap.common.ArtifactNotFoundException;
+import co.cask.cdap.internal.app.runtime.plugin.PluginNotExistsException;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.artifact.ApplicationClassInfo;
+import co.cask.cdap.proto.artifact.ApplicationClassSummary;
+import co.cask.cdap.proto.artifact.ArtifactSortOrder;
+import co.cask.cdap.proto.id.ArtifactId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.security.authorization.AuthorizationUtil;
+import co.cask.cdap.security.impersonation.EntityImpersonator;
+import co.cask.cdap.security.spi.authentication.AuthenticationContext;
+import co.cask.cdap.security.spi.authorization.AuthorizationEnforcer;
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.inject.name.Named;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * /**
+ * A class which has a delegation {@link ArtifactRepository} which has authorization enforce on the methods.
+ */
+public class AuthorizationArtifactRepository implements ArtifactRepository {
+  private static final Logger LOG = LoggerFactory.getLogger(AuthorizationArtifactRepository.class);
+
+  private final ArtifactRepository delegate;
+  private final AuthorizationEnforcer authorizationEnforcer;
+  private final AuthenticationContext authenticationContext;
+
+  @Inject
+  public AuthorizationArtifactRepository(@Named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO)
+                                           ArtifactRepository artifactRepository,
+                                         AuthorizationEnforcer authorizationEnforcer,
+                                         AuthenticationContext authenticationContext) {
+    this.delegate = artifactRepository;
+    this.authorizationEnforcer = authorizationEnforcer;
+    this.authenticationContext = authenticationContext;
+  }
+
+  @Override
+  public CloseableClassLoader createArtifactClassLoader(Location artifactLocation,
+                                                        EntityImpersonator entityImpersonator) throws IOException {
+    return delegate.createArtifactClassLoader(artifactLocation, entityImpersonator);
+  }
+
+  @Override
+  public void clear(NamespaceId namespace) throws Exception {
+    List<ArtifactSummary> artifacts = delegate.getArtifactSummaries(namespace, false);
+    for (ArtifactSummary artifactSummary : artifacts) {
+      authorizationEnforcer.enforce(namespace.artifact(artifactSummary.getName(), artifactSummary.getVersion()),
```

Comment:
One more place that clears all entities in a namespace that is related to https://github.com/caskdata/cdap/pull/9337#discussion_r132806630